### PR TITLE
Autodetect Widgettype

### PR DESCRIPTION
**Pull request by signedav**

### CODE REVIEW

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-689-a11debc1c3dd5ba679e60e0d1e8bff01e3c81b79-armv7.apk)
 - [arm64_v8a](https://download.opengis.ch/qfield/ci-builds/qfield-dev-689-a11debc1c3dd5ba679e60e0d1e8bff01e3c81b79-arm64_v8a.apk)
 - [x86](https://download.opengis.ch/qfield/ci-builds/qfield-dev-689-a11debc1c3dd5ba679e60e0d1e8bff01e3c81b79-x86.apk)
 - [x86_64](https://download.opengis.ch/qfield/ci-builds/qfield-dev-689-a11debc1c3dd5ba679e60e0d1e8bff01e3c81b79-x86_64.apk)

File: src/core/attributeformmodelbase.h [hunk 1]
```diff
@@ -83,6 +83,8 @@ class AttributeFormModelBase : public QStandardItemModel
 
     void setConstraintsValid( bool constraintsValid );
 
+    QgsEditorWidgetSetup findBest( int fieldIndex );
```
Comment (signedav) [inline src/core/attributeformmodelbase.h]: TODO: write documentation for this

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-689-87c005d096185e33dec371b8822c52da6e05a9e3-armv7.apk)
 - [arm64_v8a](https://download.opengis.ch/qfield/ci-builds/qfield-dev-689-87c005d096185e33dec371b8822c52da6e05a9e3-arm64_v8a.apk)
 - [x86](https://download.opengis.ch/qfield/ci-builds/qfield-dev-689-87c005d096185e33dec371b8822c52da6e05a9e3-x86.apk)
 - [x86_64](https://download.opengis.ch/qfield/ci-builds/qfield-dev-689-87c005d096185e33dec371b8822c52da6e05a9e3-x86_64.apk)

File: src/core/attributeformmodelbase.cpp [hunk 2]
```diff
@@ -407,6 +408,53 @@ void AttributeFormModelBase::setConstraintsValid( bool constraintsValid )
...
+    if ( fields.fieldOrigin( index ) == QgsFields::OriginProvider )
+    {
+      int providerOrigin = fields.fieldOriginIndex( index );
+      if ( !mLayer->dataProvider()->defaultValueClause( providerOrigin ).isEmpty() )
+        return setup;
+    }
+
+    //find the best one
+    //on a boolean type take "CheckBox"
+    if ( fields.at( index ).type() == QVariant::Bool )
```
Comment (m-kuhn) [inline src/core/attributeformmodelbase.cpp]: I'd probably add a single get on top

```
const QgsField field = fields.at( index );
```

Sorry for the nitpicking ;)

**Comment by m-kuhn** [comment]
Ping @gacarrillor 
is there someone from your team that could test this?

**Comment by gacarrillor** [comment]
> Ping @gacarrillor
> is there someone from your team that could test this?

Sure @m-kuhn, what exactly it does?

**Comment by m-kuhn** [comment]
Automatically use the appropriate widget types in QField if there has been no manual configuration for the widget type (e.g. when using model baker to create a project).

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-689-d62eb82426241b3afc3b0a03219d013cd4f9e195-armv7.apk)
 - [arm64_v8a](https://download.opengis.ch/qfield/ci-builds/qfield-dev-689-d62eb82426241b3afc3b0a03219d013cd4f9e195-arm64_v8a.apk)
 - [x86](https://download.opengis.ch/qfield/ci-builds/qfield-dev-689-d62eb82426241b3afc3b0a03219d013cd4f9e195-x86.apk)
 - [x86_64](https://download.opengis.ch/qfield/ci-builds/qfield-dev-689-d62eb82426241b3afc3b0a03219d013cd4f9e195-x86_64.apk)

**Comment by jhonalex06** [comment]
We tested it using Qgis Model Baker, and when we generate the project we save it in two modes: the first one without using QFieldSync and the second one using QFieldSync. In both modes the result is good, all widgets work correctly.

Thanks

**Pull request merged**